### PR TITLE
dmd: update to 2.094.1

### PR DESCRIPTION
**Pull request by aurieh**
Also adds a new package, `dmd2.091`, which is required to bootstrap `2.094.1`, else the build fails with

```
Error: pointer cast from dmd.mtype.TypeTuple to immutable(void*)** is not supported at compile time
```
in `dmd/mtype.d`.

### CODE REVIEW

**Comment by aurieh** [comment]
`2.094.1` breaks on `x86_64-musl` due to missing `vfork` (?!), not yet sure why.

File: srcpkgs/dmd2.091/template
Comment (sgn) [inline srcpkgs/dmd2.091/template:8]: Let's not split `libphobos2.091` out.

File: srcpkgs/dmd/template
Comment (ericonr) [inline srcpkgs/dmd/template:24]: Can we replace any of these in the bootstrap chain?

File: srcpkgs/dmd2.091/template
Comment (sgn) [inline srcpkgs/dmd2.091/template:82]: And I think we shouldn't ship documentation either. This package is only used to build latest `dmd`, no?

File: srcpkgs/dmd2.091/template
Comment (ericonr) [inline srcpkgs/dmd2.091/template:1]: If it's only going to be used for bootstrap, could we simplify the package with fewer subpackages and rm'ing everything that's not important?

File: srcpkgs/dmd/template
Comment (sgn) [inline srcpkgs/dmd/template:42]:
```sh
if [ "$XBPS_TARGET_LIBC" = musl ]; then
...
fi
```

Comment (sgn) [reply srcpkgs/dmd/template:42]: But, musl has `vfork(2)` from like `0.8.4`??

**Review by sgn — CHANGES_REQUESTED** [review]

File: srcpkgs/dmd/files/musl.patch
Comment (ericonr) [inline srcpkgs/dmd/files/musl.patch:18]: Why is a return here? Using a `return` feels very wrong.... I don't think I would revert this line.

`vfork` is available on musl, they are probably failing to `#define _BSD_SOURCE` at the top of the file, or whatever the D equivalent is.

**Review by ericonr — COMMENTED** [review]
Remember to squash the commits as well.

File: srcpkgs/dmd/template
Comment (aurieh) [reply srcpkgs/dmd/template:24]: No. `dmd-bootstrap` is needed to build `2.081` (obviously), `2.081` is needed to build `2.091`  (and probably older versions, not sure what range is but it doesn't matter because...) `2.091` is needed to build anything past `2.092.1` (thanks to a  change in `mtype.d`).

File: srcpkgs/dmd2.091/template
Comment (aurieh) [reply srcpkgs/dmd2.091/template:1]: Yes, as pointed out by @sgn, I should probably just get rid of everything non-essential. I should probably trim the `dmd2.081`  package as well, although in a different PR.

File: srcpkgs/dmd/files/musl.patch
Comment (aurieh) [reply srcpkgs/dmd/files/musl.patch:18]: D does not use C headers. I tried to track down the declaration of `vfork` in glibc druntime, which led me to https://github.com/dlang/druntime/blob/master/src/core/sys/posix/unistd.d#L2451 but declaring `vfork` here https://github.com/dlang/druntime/blob/master/src/core/sys/posix/unistd.d#L2479 (in the bootstrap compiler's druntime, of course) did not help, which leads me to believe the musl block is never evaluated thanks to D toolchain's poor support of anything other than x86_64 glibc.

File: srcpkgs/dmd/template
Comment (aurieh) [reply srcpkgs/dmd/template:42]: I explained the reason for this in https://github.com/void-linux/void-packages/pull/25878#discussion_r511674166

File: srcpkgs/dmd/files/musl.patch
Comment (aurieh) [reply srcpkgs/dmd/files/musl.patch:18]: And yes, good point, I will not revert that line. I just sort of took a hammer approach with the revert, took the whole commit out.

Comment (ericonr) [reply srcpkgs/dmd/files/musl.patch:18]: Gah, that's terrible.

> (in the bootstrap compiler's druntime, of course)

Are you sure that's correct? The bootstrap compiler should be building the new compiler with its own library and headers, right? Could be interesting to see what happens if you patch it in `dmd`.

Comment (ericonr) [reply srcpkgs/dmd/files/musl.patch:18]: Also, do you know how they detect musl?

Comment (aurieh) [reply srcpkgs/dmd/files/musl.patch:18]: >Are you sure that's correct?

I'm relatively certain, but I'll try patching `dmd`'s `druntime` anyways. Might work, who knows.

>Also, do you know how they detect musl?

Nope, no idea. I'll look into this tomorrow.

Comment (ericonr) [reply srcpkgs/dmd/files/musl.patch:18]: > Nope, no idea. I'll look into this tomorrow.

It's not urgent, I was mostly wondering.

Also, moving to using `vfork` when it's not present for all headers is a big bug on their part.

Comment (aurieh) [reply srcpkgs/dmd/files/musl.patch:18]: Tested patching `dmd`'s `druntime`, didn't help. Also tested adding a message pragma to the `version (CRuntime_Musl)` test in druntime - it did not log anything, which confirms that the branch is, indeed, never evaluated. Honestly not sure how this thing even builds (or works) at all.

Comment (ericonr) [reply srcpkgs/dmd/files/musl.patch:18]: Ok, no problem. So just patch it to `fork()` for now, unless you want to explore deep into the build system :)

Comment (aurieh) [reply srcpkgs/dmd/files/musl.patch:18]: Had a shot at tracking down where `CRuntime_Musl` is defined, couldn't find it; the dmd build toolchain (the entire codebase, really) is extremely convoluted. What I *do* know is that `CRuntime_Musl` is set when building both `dmd2.091` and `dmd` (`2.094.1`), made sure using a message pragma. Not sure why properly declaring `vfork()` doesn't help, it's really strange.

File: srcpkgs/dmd2.091/template
Comment (ericonr) [inline srcpkgs/dmd2.091/template:30]: No issue with using `i686*`, unless i686-musl is also unsupported by upstream.

File: srcpkgs/dmd2.091/template
Comment (ericonr) [inline srcpkgs/dmd2.091/template:35]: Could this be `rmdir`?

File: srcpkgs/dmd2.091/template
Comment (ericonr) [inline srcpkgs/dmd2.091/template:7]: Shouldn't `dmd` go in hostmakedepends? I see this is how the previous template was doing it, but still.

Comment (sgn) [reply srcpkgs/dmd2.091/template:7]: We could set this package as `nocross`, too. No point on building this package for cross.

File: srcpkgs/dmd2.091/template
Comment (aurieh) [reply srcpkgs/dmd2.091/template:30]: I do not believe i686-musl is supported upstream. I couldn't get it to build last I tried. I'll have another go with the new bootstrap chain.

File: srcpkgs/dmd2.091/template
Comment (aurieh) [reply srcpkgs/dmd2.091/template:35]: I'm not experienced enough with `xbps-src` to tell. Are the directories supposed to be empty, even after a failed build?

Comment (ericonr) [reply srcpkgs/dmd2.091/template:35]: If they are empty directories, it should be ok.

Comment (aurieh) [reply srcpkgs/dmd2.091/template:35]: Interesting, apparently the dirs don't even exist after `do_extract`. I wonder why that `rm` is in the original template.

File: srcpkgs/dmd2.091/template
Comment (aurieh) [reply srcpkgs/dmd2.091/template:31]: Tried building on `i686-musl`, `dmd2.081` fails because its druntime is missing `pthread_condattr_t` even though it's defined for musl in https://github.com/dlang/druntime/blob/fa3eeed29b5d892da921c585e6526635475059a4/src/core/sys/posix/sys/types.d#L852 *and* the version block is evaluated.

**Comment by aurieh** [comment]
@ericonr @sgn anything else I need to get done before this is ready to be merged?

File: srcpkgs/dmd2.091/template
Comment (ericonr) [reply srcpkgs/dmd2.091/template:31]: Wow, that sucks. Thanks for testing!

**Comment by ericonr** [comment]
I don't think there's anything, besides possibly cleaning up the body of the commit messages (they have some repeated statements).

If it isn't merged by tomorrow, I will try out something built with it and do so.

**Comment by aurieh** [comment]
Cleaned up the commit messages.

**Comment by ericonr** [comment]
Should https://github.com/void-linux/void-packages/pull/25879 be built with the new dmd?

**Comment by aurieh** [comment]
It builds with `2.091.0` but, just in case, I think it'd be safer to merge this first, then #25879.

**Comment by ericonr** [comment]
Ok, thank you

**Comment by ericonr** [comment]
I'm testing the build here to see if it behaves in my environment.

Could you report our `vfork` issues? Ideally they would fix it in a new release; the issues with Musl detection and such should also be noted.

Bwt, I would backport the `return` -> `_exit` commit to all versions. If it fails to exec `ld`, the whole process state will be kinda fucked.

**Comment by ericonr** [comment]
Building on musl is hanging for me... I will try using less parallel jobs.

**Comment by ericonr** [comment]
It might be necessary to add `disable_parallel_builds=yes` to the template, methinks. Could be conditional on it being musl.

**Comment by aurieh** [comment]
>Could you report our vfork issues? Ideally they would fix it in a new release; the issues with Musl detection and such should also be noted.

Will do, although I don't believe musl detection is broken. I double-checked everything and it executes the musl version blocks just fine at compile time. At some point I was told by a dmd maintainer for a different distro that the issues with alternative libcs trace back to the bootstrap compiler (`dmd-bootstrap` here), but I'm not sure what exactly the issue is.

>Bwt, I would backport the return -> _exit commit to all versions. If it fails to exec ld, the whole process state will be kinda fucked.

Sounds good to me.

>Building on musl is hanging for me... I will try using less parallel jobs.

Had that happen as a one-off thing on `x86_64`, glibc I believe, running 16 jobs. Although the place where it hung for me was a non-parallel invocation of the newly-built dmd to build druntime (`?`).

>It might be necessary to add disable_parallel_builds=yes to the template, methinks. Could be conditional on it being musl.

Due to aforementioned uncertainty, it might be a good idea to add it for both glibc and musl. Then again, not sure if it's parallel builds that cause this though.

**Comment by aurieh** [comment]
Backported the patch. Now all that's left is to determine whether parallel jobs are what occasionally break the build. Can you consistently reproduce it?

**Comment by ericonr** [comment]
>  Can you consistently reproduce it?

Not really :/

**Comment by ericonr** [comment]
I was going to try and build it once more, but building dmd2.091 just hung for 1h :/

**Comment by aurieh** [comment]
Last I tried, I was unable to reproduce it with parallel builds disabled, so that might very well be the solution. How many jobs did you use?

**Comment by ericonr** [comment]
Yeah, I had 13 jobs. Will try again with them disabled.

**Comment by ericonr** [comment]
Still hanging... I'm not sure what to do :/

**Comment by ericonr** [comment]
@aurieh can you add `disable_parallel_build=yes` to the templates and see if they pass CI?

**Comment by aurieh** [comment]
@ericonr They do, but this doesn't seem to be a consistently reproducible problem.

**Comment by ericonr** [comment]
@aurieh can you move this to 2.095.1 and we try the dance again?

**Comment by aurieh** [comment]
@ericonr done! Sorry for taking so long, life got in the way of many things.

**Pull request closed without merging**
